service/history/queues/iterator.go

```diff
@@ -33,7 +33,8 @@ import (
 
 type (
 	Iterator interface {
-		collection.Iterator[tasks.Task]
+		HasNext(readerID int32) bool
+		Next(readerID int32) (tasks.Task, error)
```

Comment:
I want iterators to be able to freely move between slices & queue readers, instead of bind to a specific slice or reader. Existing implementation is also based on that idea. 

So readerID here can be regarded as a context for loading operations. The iterator implementation knows that if readerID changes, a new underlying pagingIterator is needed for talking to persistence.